### DISCUSSION TITLE
Force undo changes to a respository

### DISCUSSION THREAD

**Discussion by eschan145** [community/community#30012] [Other Feature Feedback, Questions, & Ideas]
I'm completely new to using Git, and before I only used online GitHub. In one of my latest commits, I had a typo, so I tried fixing it using Git GUI. But I accidentally deleted like 200 commits when I only had 300. How can I force undo?

**Comment by airtower-luna** [comment]
First, if you still have the history before the mistake in a remote branch (e.g. in a repository on GitHub), you can recover by *resetting* your local branch to that, or creating a new branch based on that remote branch. That's the easy case.

Otherwise, if the lost history was local-only: Most likely the "lost" commits are still somewhere in the repository data. The tricky thing is finding them, you need to figure out the commit IDs. You can use `git reflog` to see at which commits and branches your repository was in the past. The commit right before you tried to fix that one commit should be listed in there. Once you have the right ID you can start a new branch from there.

[Git Internals - Maintenance and Data Recovery](https://git-scm.com/book/en/v2/Git-Internals-Maintenance-and-Data-Recovery) has a detailed description.

**Reply by eschan145** [reply]
The history is on GitHub, but I'm trying to change it back to a local respository

**Reply by airtower-luna** [reply]
If you still have the history on GitHub, you can fetch it from there (I'm going to assume the remote for your GitHub repository is named `origin`):

```
git fetch origin
```

`fetch` only makes sure your local repository *knows* the history from GitHub, it doesn't make any changes to your files. You can then *check out* the branch that has the history you want. It's easiest if you create a new branch from the remote branch. For example if the branch in your GitHub repository is named `mouse`, and you want to call the new one `my-mouse`:

```
git checkout -b my-mouse origin/mouse
```

Then you have a new local branch with the history you want. You could also use `git reset` to force your current local branch to the state of a remote one, but it's way easier to make mistakes with that. :sweat_smile: Creating a new branch is safe, it doesn't change any of your existing branches. And if/when you're sure you don't want some of the old branches you can still delete them. Or rename the new one. :slightly_smiling_face: 

[Git Basics - Working with Remotes](https://git-scm.com/book/en/v2/Git-Basics-Working-with-Remotes) explains the "remote" thing in a lot more detail. In general I recommend you read (at least) [chapters 2 and 3 of the Git Book](https://git-scm.com/book/en/v2), it's a pretty good introduction.

**Reply by eschan145** [reply]
The branch doesn't appear in the repo

**Reply by airtower-luna** [reply]
How do you mean? You wrote above that the history is on GitHub.
